subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java

```diff
@@ -81,8 +89,7 @@ public void execute(final TaskInternal task, TaskStateInternal state, TaskExecut
                         cacheable = true;
                         TaskArtifactState taskState = context.getTaskArtifactState();
                         try {
-                            cacheKey = taskState.calculateCacheKey();
-                            LOGGER.info("Cache key for {} is {}", task, cacheKey);
+                            cacheKey = cacheKeyCalculator.calculate(taskState.getCurrentExecution(), task);
```

Comment:
I think the cache key calculator should be wired in through the `TaskArtifactStateRepository` and in `TaskArtifactState`, we could save the result of calculating the `BuildCacheKey`.  That way there would be fewer services required here and `SkipCachedTaskExecutor` would be doing less.

This might make wiring in the listener different since you wouldn't know how the cache key was about to be used.  Maybe that would let us build a better listener that compares the before and after keys at a deeper level?